### PR TITLE
[ENG-2204]  Branded Registries where the EGAP schema is showing for all providers when you have the flag set, not just the provider who has it as an acceptable schema

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

This should fix the issue where the EGAP schema was showing up for every superuser withour effecting non-super users.

## Changes

- simple logic change

## QA Notes

  - Does this change require a data migration? If so, what data will we migrate?
No
  - What is the level of risk?
Low
    - Any permissions code touched?
Kinda?
    - Is this an additive or subtractive change, other?
other
  - How can QA verify? (Through UI, API, AdminApp or AdminAdminApp?)

Aside from testing the bug you explicitly found, please check the traditional registration workflow (through the project's Registrations tab) and verify that nobody, including EGAP Admins, can submit an egap registration through that path.

## Documentation

Bug fix, no docs.

## Side Effects

EGAP Schema may have to be made `visible` to be usable.

## Ticket

https://openscience.atlassian.net/browse/ENG-2204

### CODE REVIEW

File: api_tests/providers/registrations/views/test_registration_provider_schemas.py [hunk 1]
```diff
@@ -71,6 +71,9 @@ def provider_with_v2_prereg_only(self, schema):
     def provider_with_egap_only(self, egap_schema):
         provider = RegistrationProviderFactory()
         update_provider_auth_groups()
+        egap_schema.visible = True
```
Comment (brianjgeiger) [inline api_tests/providers/registrations/views/test_registration_provider_schemas.py:74]: Drive-by: Hmm. I thought we ended up changing this view so it did not require the schema to be visible. Maybe I thought about it, but it didn't seem worth the change. For the `/v2/schemas/registrations/` endpoint, I would expect that we would need the schema to be visible, but for here, it shouldn't need that.

Comment (Johnetordoff) [reply api_tests/providers/registrations/views/test_registration_provider_schemas.py:74]: That endpoint only delivers visible schemas, IMO it's better just to make the egap schema visible then to make the special casing special-er. It won't turn up any place import because it's only associated with the EGAP provider.

Comment (brianjgeiger) [reply api_tests/providers/registrations/views/test_registration_provider_schemas.py:74]: The `visible` attribute only makes sense for `/v2/schemas/registrations`. We have another method of choosing what schemas are visible in this view (i.e. explicitly choosing them for the provider). When we end up adding other providers, we're going to want them not to show in `/v2/schemas/registrations`, because they won't be for the OSF provider using the existing workflow. They'll only be for the new providers. 

We should make this endpoint show regardless of visibility, but keep the other one working as it is.

File: osf/models/metaschema.py [hunk 2]
```diff
@@ -59,7 +59,7 @@ def get_latest_versions(self, request=None):
         :param request: the request object needed for waffling
         :return: queryset
         """
-        queryset = self.filter(visible=True).order_by('name', '-schema_version').distinct('name')
+        queryset = self.order_by('name', '-schema_version').distinct('name')
```
Comment (brianjgeiger) [inline osf/models/metaschema.py:62]: There are three places where this is used.

1. api.schemas.views:RegistrationSchemaList::get_default_queryset()
2. website.project.views.drafts::get_metaschemas()
3. api.providers.views:RegistrationProviderSchemaList::get_default_queryset()

The first two expect only visible items to return. The third expects to get them regardless of visibility. So the thing that makes the most sense to me is to have a parameter to show invisible that defaults to `False`, and set it in `api.providers.views:RegistrationProviderSchemaList::get_default_queryset()` to `True`.

The reason for the difference is that the first two are related to the traditional method of making a registration, which is to go through a node's registration tab and make a registration from there. That defaults to OSF as the provider, and we haven't updated that to use the new `api.providers.views:RegistrationProviderSchemaList`. Therefore, it needs to have all of the latest, active, visible schemas and none of the rest. 

Whereas anything using the new `api.providers.views:RegistrationProviderSchemaList` needs to have whatever schemas are explicitly defined in the admin app for use. So, when we get provider Foo with schema Bar that is only used by Foo, we set that to be invisible to OSF, but we add them with the admin app on that provider. It shows in the new view but not the older views. Does that make sense?

File: osf/models/metaschema.py [hunk 3]
```diff
@@ -39,7 +39,7 @@ def allow_egap_admins(queryset, request):
     if hasattr(request, 'user') and waffle.flag_is_active(request, EGAP_ADMINS):
         return queryset | RegistrationSchema.objects.filter(name='EGAP Registration').distinct('name')
     else:
-        return queryset
+        return queryset.exclude(name='EGAP Registration')
```
Comment (brianjgeiger) [inline osf/models/metaschema.py:42]: This will work, but I think the previous way you had it, with the change below, would be more correct.

File: api/providers/views.py [hunk 4]
```diff
@@ -615,7 +616,8 @@ def get_provider(self):
         return RegistrationProvider.objects.get(_id=self.kwargs['provider_id'])
 
     def get_default_queryset(self):
-        return self.get_provider().schemas.get_latest_versions(request=self.request).filter(active=True)
+        queryset = self.get_provider().schemas.get_latest_versions().filter(active=True, visible=True)
```
Comment (brianjgeiger) [inline api/providers/views.py:619]: This should not be filtering `visible=True`

File: osf/models/metaschema.py (hunk 3 above)
Comment (brianjgeiger) [reply osf/models/metaschema.py:42]: Thinking more, I don't think this is going to work. I don't see how it's going to prevent other providers from showing the egap registration form when the user has permissions to see that form.

File: api_tests/providers/registrations/views/test_registration_provider_schemas.py [hunk 5]
```diff
@@ -100,9 +102,10 @@ def test_registration_provider_with_schema(
...
 
-        assert len(data) == 1
-        assert data[0]['id'] == schema._id
-        assert data[0]['attributes']['name'] == schema.name
+        assert len(data) == 2
+        assert schema._id in [item['id'] for item in data]
+        assert invisible_schema._id in [item['id'] for item in data]
+        assert schema.name in [item['attributes']['name'] for item in data]
 
         res = app.get(url_with_v2_prereg_only, auth=user.auth)
```
Comment (brianjgeiger) [inline api_tests/providers/registrations/views/test_registration_provider_schemas.py:110]: The user for the requests in this test needs to be an `egap_admin`. That should bring about the right conditions for the fix that needs to go in.

**Review by brianjgeiger — APPROVED** [review]

File: api_tests/schemas/views/test_registration_schemas_list.py [hunk 6]
```diff
@@ -71,7 +71,3 @@ def test_schemas_list_crud(self, app, url, user, egap_admin, factory_request):
         res = app.get(url, auth=user.auth)
         assert res.status_code == 200
         assert not [data for data in res.json['data'] if data['attributes']['name'] == 'EGAP Registration']
-
-        res = app.get(url, auth=egap_admin.auth)
-        assert res.status_code == 200
-        assert [data for data in res.json['data'] if data['attributes']['name'] == 'EGAP Registration']
```
Comment (mfraezz) [inline api_tests/schemas/views/test_registration_schemas_list.py:77]: Minor: Would it be better to `assert not` rather than to not `assert` here?

**Review by mfraezz — APPROVED** [review]
Couple questions, but no blockers. Pass complete :octocat:

File: osf/models/metaschema.py [hunk 7]
```diff
@@ -36,11 +36,9 @@ def allow_egap_admins(queryset, request):
     Allows egap admins to see EGAP registrations as visible, should be deleted when when the EGAP registry goes
```
[the remark this answers is not in the document]
Comment (brianjgeiger) [reply osf/models/metaschema.py:36]: No, we still need it for the EGAP Registry as things stand. It just no longer affects the traditional registration workflow.
